### PR TITLE
Show Read Receipts for grouped event summaries

### PR DESCRIPTION
**Pull request by justjanne**
Type: Enhancement
Closes: https://github.com/vector-im/element-web/issues/21717

---
Here's what your changelog entry will look like:

## ✨ Features
 * Show Read Receipts for grouped event summaries ([\#8446](https://github.com/matrix-org/matrix-react-sdk/pull/8446)). Fixes vector-im/element-web#21717. Contributed by @justjanne.

### CODE REVIEW

**Comment by t3chguy** [comment]
Can we get a screenshot please, this needs design input IMO given the expand interaction for the ELS will be moving

**Comment by t3chguy** [comment]
Something looks off about the vertical alignment in the collapsed state

**Review by turt2live — COMMENTED** [review]
code review isn't really practical on this without design review

**Comment by justjanne** [comment]
> code review isn't really practical on this without design review

Do we have a figma design for event summaries? I can’t find anything that’d specify metrics. So far, everything seems to be aligned ad-hoc. It’d be great if we had proper specs for expanded/collapsed in different layouts, what spacing we should use, how it’d shift if the summary wraps (or is short enough not to wrap).

**Comment by t3chguy** [comment]
They predate the use of figma in the company and haven't been redesigned since

**Comment by niquewoodhouse** [comment]
@justjanne how do I see a build of this to review it? I can't work out what's being asked to review.

Is it just making the read receipts by default visible next to grouped events like "John and 93 others joined, John and 90 others left."?

**Comment by justjanne** [comment]
> Is it just making the read receipts by default visible next to grouped events like "John and 93 others joined, John and 90 others left."?

Yes, that’s exactly it. Including how the read receipts are aligned with the text and the other avatars on the left.

**Comment by t3chguy** [comment]
(fixing Netlify)

**Review by niquewoodhouse — COMMENTED** [review]
Have looked and in terms of actual UI design its ok, I just have some comments, but I'm not sure in terms of ux or product if this is helpful enough to people to be worth making the timeline slightly harder for everyone.

I've tried to put my comments into themes.

**UI**

- [x] There's no hover state for event summaries that lets me know the receipts are referring to them.  

https://user-images.githubusercontent.com/1587166/166479571-38245c01-f6fb-4ea9-9f98-6b31f934462c.mov

- [ ] Can expand be the same size as the usual copy in the timeline?  It looks 1px larger than a standard text message.

- [ ] There's an alignment issue with the expand buttons. Because this proposal can be have varying widths for the read receipts, it can push the buttons to appear at different places which makes the timeline more cluttered and harder to read

image


**Trust**

In this example, who are these people?  

image

Have these 21 people seen every event that's in the summary? Read receipts is how Element tells me people have read something, so it feels like its a point of trust, if I can't be sure what it means or it's mislead me, I lose a bit of trust in the product.

**Visual space**

Adding this information to be viewable by default makes the timeline a bit busier, which makes it a tiny bit harder to read.  I don't really understand how important or useful seeing this by default is when it means the timeline is harder and it's not clear what it means.  Do we have any insights around why this is important or helpful?

**Review by niquewoodhouse — CHANGES_REQUESTED** [review]
Thank you for addressing those concerns.  

- [ ] Adjust positions The expand is now the same size as the other text but it looks like its just slightly too low, please align expand so it starts on the same baseline as the text to the left.

image

**Comment by justjanne** [comment]
I’ve aligned it now as closely as I can, as they’re positioned slightly differently in some situations (I think it only occurs with the bubble theme) a misalignment of 0.5 pixels is unavoidable, but it shouldn’t be noticeable.

**Review by niquewoodhouse — APPROVED** [review]
This works/looks as expected for me, thanks for that.

My only concern remains trust and meeting expectations.  The counter of readers visible for a grouped event summary reflects everyone who has read all the items within the group.  I think that's correct, and is better than saying the wrong thing, but I think we'll find out more in practice. 

Thanks

File: src/components/structures/MessagePanel.tsx
Comment (t3chguy) [inline src/components/structures/MessagePanel.tsx:876]: this seems to only show the RR for the last event? isn't that more confusing that the union?

Comment (justjanne) [reply src/components/structures/MessagePanel.tsx:876]: If we've got messages A B and C in a group, and a user acknowledges message C, then we know the user has read all of the messages.

Using the RRs of the last event ensures that we only show RRs for the group if a user has acknowledged all the messages in the group. The union would also show RRs of users who had only read some of the messages in the group

Comment (t3chguy) [reply src/components/structures/MessagePanel.tsx:881]: Yes but imagine events:

Message 1
Message 2
State 1
State 2
State 3

(State events being in a group)

If a user's RR is set on a Message (1/2) and then they send a read receipt for State (1/2) then their RR will vanish, which will look broken IMO

Comment (justjanne) [reply src/components/structures/MessagePanel.tsx:881]: That vanishing is already what happens today. But if we show a union of read receipts, we’d risk the user not understanding what has been read and what hasn’t.

Should product take a look at this and make the decision, as nique suggested last week?

**Review by nadonomy — CHANGES_REQUESTED** [review]
So after looking at this in more detail, we've decided against merging this. The feature seems attractive upon first glance but in practise creates more problems than it solves:

1. It creates ambiguity over what the expand button does (does it expand the event summary? read receipts?).
2. It creates ambiguity over what the read receipts are summarising (either they neglect partial event reads which some users will find unexpected, or it includes them which would be misleading).
3. It reduces the signal of read receipts for higher value events, like user written messages.
4. It reduces the legibility of the timeline by adding extraneous elements to it.
5. It creates unnecessary complex requirements for how we might stylise events more distinctly to messages in the timeline (where we made recent learnings on promoting legibility, from implementing message bubbles).

On aggregate, we think it detracts more value than it adds so don't think it's worth the tradeoff to merge this into main, and don't see a high enough user need or demand to suggest swallowing the debt of moving it to labs.

**Comment by justjanne** [comment]
Should we close https://github.com/vector-im/element-web/issues/21717 as WONTFIX then accordingly?

**Comment by nadonomy** [comment]
@justjanne I'm not sure we use WONTFIX labels ([triage process](https://github.com/vector-im/element-meta/wiki/Triage-process), [issue labels](https://github.com/vector-im/element-meta/wiki/Issue-labelling)) but will defer to the repo maintainers on that one.

With that issue specifically I'd advocate to do one of:

1. Close the issue as the suggested solution isn't tractable.
2. Keep it open to gather more discussion, editing to (1) make clearer it's one suggested solution with now known flaws (2) pivot discussion to the user need so we can find the best solution.

**Comment by JokerGermany** [comment]
Thats a pitty, i think it's a pain having to expand it first to look who read what..

**Pull request closed without merging**
